exporter/honeycombexporter/honeycomb.go

```diff
@@ -16,91 +16,287 @@ package honeycombexporter
-	index := 0
-	for key, val := range attributes {
-		result[index] = core.Key(key).String(val)
-		index++
+// sendSpanLinks gets the list of links associated with this span and sends them as
+// separate events to Honeycomb, with a span type "link".
+func (e *honeycombExporter) sendSpanLinks(span *tracepb.Span) {
+	links := span.GetLinks()
+
+	if links == nil {
+		return
+	}
+
+	for _, l := range links.GetLink() {
+		ev := e.builder.NewEvent()
+		ev.Add(link{
+			TraceID:     getHoneycombTraceID(span.GetTraceId()),
+			ParentID:    getHoneycombSpanID(span.GetSpanId()),
+			LinkTraceID: getHoneycombTraceID(l.GetTraceId()),
+			LinkSpanID:  getHoneycombSpanID(l.GetSpanId()),
+			SpanType:    "link",
+			RefType:     spanRefType(l.GetType()),
+		})
+		attrs := spanAttributesToMap(l.GetAttributes())
+		for k, v := range attrs {
+			ev.AddField(k, v)
+		}
+		if err := ev.Send(); err != nil {
+			e.onError(err)
+		}
+	}
+}
+
+// sendMessageEvents gets the list of timeevents from the span and sends them as
+// separate events to Honeycomb, with a span type "span_event".
+func (e *honeycombExporter) sendMessageEvents(td consumerdata.TraceData, span *tracepb.Span, traceFields map[string]interface{}) {
+	timeEvents := span.GetTimeEvents()
+	if timeEvents == nil {
+		return
+	}
+
+	for _, event := range timeEvents.TimeEvent {
+		annotation := event.GetAnnotation()
+		if annotation == nil {
+			continue
+		}
+
+		ts := timestampToTime(event.GetTime())
+		name := annotation.GetDescription().GetValue()
+		attrs := spanAttributesToMap(annotation.GetAttributes())
+
+		// treat trace level fields as underlays with same keyed span attributes taking precedence.
+		ev := e.builder.NewEvent()
+		for k, v := range traceFields {
+			ev.AddField(k, v)
+		}
+		if span.Resource != nil && span.Resource.Labels != nil {
+			resourceType := span.Resource.GetType()
+			if len(resourceType) != 0 {
+				ev.AddField("resource_type", resourceType)
+			}
+			for k, v := range span.Resource.Labels {
+				ev.AddField(k, v)
+			}
+		}
+		if len(span.GetParentSpanId()) == 0 || hasRemoteParent(span) {
+			if td.Node != nil {
+				for k, v := range td.Node.Attributes {
+					ev.AddField(k, v)
+				}
+			}
+		}
+		for k, v := range attrs {
+			ev.AddField(k, v)
+		}
+		ev.Timestamp = ts
+		ev.Add(spanEvent{
+			Name:       name,
+			TraceID:    getHoneycombTraceID(span.GetTraceId()),
+			ParentID:   getHoneycombSpanID(span.GetSpanId()),
+			ParentName: truncatableStringAsString(span.GetName()),
+			SpanType:   "span_event",
+		})
+		if err := ev.Send(); err != nil {
+			e.onError(err)
+		}
 	}
-	return result
 }
 
-func (e *HoneycombExporter) Shutdown(context.Context) error {
-	e.exporter.Close()
+// Shutdown takes care of any cleanup tasks that need to be carried out. In
+// this case, we close the honeycomb sdk which flushes any events still in the
+// queue and closes any open channels between queues.
+func (e *honeycombExporter) Shutdown(context.Context) error {
+	libhoney.Close()
```

Comment:
Is this safe to call multiple times? That can happen if there are multiple exporters configured.

Reply:
Yup, thanks for checking - it'll be a noop if it's already closed.

Reply:
> Please resolve the conflicts.

Done. :+1: